adr/0003-multiple-templates-per-component.md

```diff
@@ -0,0 +1,52 @@
+# 1. Allow multiple templates 
+
+Date: 2021-08-30
+
+## Status
+
+Proposed.
+
+## Context
+
+As components become larger (for example, because you are implementing a whole page), it becomes
+useful to be able to extract sections of the view to a different file. ActionView has
+partials, and ViewComponent lacks a similar mechanism. 
+
+ActionView partials have the problem that their interface is not introspectable. Data
+may be passed into the partial via ivars or locals, and it is impossible to know
+which without actually opening up the file. Additionally, partials are globally 
+invocable, thus making it difficult to detect if a given partial is in use or not,
+and who are its users.
+
+An option would be to extract another Component for the extracted section, but that
+has the following drawbacks:
+
+1. It creates a new public class.
+2. If new component is invoked repeatedly (eg, for a list of items), this creates
```

Comment:
I posted something here: https://github.com/github/view_component/pull/451#issuecomment-762516144 . I'll come back with better numbers. I don't think lambdas are an option since they don't allow ERB.

Reply:
I have created a [benchmark](https://github.com/fsateler/view_component/commit/54a8ba44)

The result is:

```
Warming up --------------------------------------
               slot:   139.000  i/100ms
            partial:   299.000  i/100ms
Calculating -------------------------------------
               slot:      1.536k (± 8.9%) i/s -     15.290k in  10.038060s
            partial:      2.935k (± 6.8%) i/s -     29.302k in  10.032277s

Comparison:
            partial::     2935.2 i/s
               slot::     1536.0 i/s - 1.91x  (± 0.00) slower
```

Increasing the number of objects to iterate over to 10.000:

```
Warming up --------------------------------------
               slot:     1.000  i/100ms
            partial:     1.000  i/100ms
Calculating -------------------------------------
               slot:      1.431  (± 0.0%) i/s -     15.000  in  10.580122s
            partial:      3.183  (± 0.0%) i/s -     33.000  in  10.430793s

Comparison:
            partial::        3.2 i/s
               slot::        1.4 i/s - 2.22x  (± 0.00) slower
```

This is with ruby 2.7.3. results are similar with 3.0.1.